File: src/main/java/com/moovie/backend/domain/usecase/movies/DiscoverMovies.java
package com.moovie.backend.domain.usecase.movies;

import com.moovie.backend.domain.repository.MoviesRepository;
import com.moovie.backend.domain.model.MovieListing;
import org.springframework.stereotype.Service;

@Service
public class DiscoverMovies {

    private final MoviesRepository moviesRepository;

    public DiscoverMovies(MoviesRepository moviesRepository) {
        this.moviesRepository = moviesRepository;
    }

    public MovieListing invoke(int page, String year, String releaseDateGte, String releaseDateLte,
                                String sortBy, String withGenres, String withOriginalLanguage,
                                String withOriginCountry, String voteCountGte) {
        return moviesRepository.discoverMovies(page, year, releaseDateGte, releaseDateLte,
                sortBy, withGenres, withOriginalLanguage, withOriginCountry, voteCountGte);
    }
}

File: src/main/kotlin/org/mobyle/domain/usecase/movies/DiscoverMovies.kt
package org.mobyle.domain.usecase.movies

import kotlinx.coroutines.runBlocking
import org.mobyle.domain.repository.MoviesRepository
import org.mobyle.model.MovieListing

class DiscoverMovies(private val repository: MoviesRepository) {
    operator fun invoke(
        page: Int,
        year: Int? = null,
        releaseDateGte: String? = null,
        releaseDateLte: String? = null,
        sortBy: String? = null,
        genres: String? = null,
        language: String? = null,
        country: String? = null,
        voteCountGte: Int? = null
    ): MovieListing = runBlocking {
        repository.discoverMovies(
            page = page,
            year = year,
            releaseDateGte = releaseDateGte,
            releaseDateLte = releaseDateLte,
            sortBy = sortBy,
            genres = genres,
            language = language,
            country = country,
            voteCountGte = voteCountGte
        )
    }
}
